TLS Audit

rija opened the pull request:
This PR implements the changes that comes out from auditing our TLS setup for issue #818:

* [x] Run deployed web site through Qualys SSL Labs certificate check
* [x] Verify connection upgrade, that "http://" redirect to "https://" and that [HSTS](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Strict-Transport-Security) is enabled and set up correctly
* [x] Disable old versions of TLS and enable TLS  v1.3
* [x] Create an automated smoke test that fails if expiration of a certificate is less than 10 days
* [x] Update docs, add a ``docs/TLS.md`` to document TLS  usage, setup and security considerations in GigaDB 

# Changes

* Added an operations script ``ops/scripts/cert_age.sh`` that return how old the certificate is
* Added an automated test ``tests/check_cert.sh`` set to fail if the Let's Encrypt certificate for staging and live are close to their expiration date and added it to ``tests/all_and_coverage``
* Updated Nginx configuration to pass the Qualys SSL Labs automated audit with a A+ grade
* Add documentation to explain how TLS certificates are set up and used in our project
* Updated the security documentation with a section about our TLS setup
* Updated the ``before_script`` section of ``.gitlab-ci.yml`` to make useful tools ``jq`` and ``curl`` available to docker dind's shell (and used in ``check_cert.sh``) in gitlab jobs scripts' steps

rija committed:
Enable TLS 1.3 and disable TLS 1 and 1.1

the older protocols have weaker cyphers and have more vulnerabilities

Disabling them reduce our TLS-related surface attack

rija committed:
Remove redundant HSTS declaration

it needs to appear only once

rija committed:
Add script to check cert expiry and automated tests for production

cert_age.sh will calculate whether the cert for the domain passed as an argument
will expire soon (10 days or less) or not

cert_check will use the script to test the REMOTE_HOSTNAME on staging and live are not
near expiry date.

We then make sure the cert check is run as part of the automated tests in CI

rija committed:
Add documentation of the use of TLS certificate in our project

rija committed:
Add documentation on security of our TLS setup

rija committed:
Add small section about openssl in security docs

rija committed:
use CI_PROJECT_PATH to access the GitLab API

because PROJECT_VARIABLES_URL is only available on developer machine

pli888 reviewed (approved):
The document is clear and informative. I have also checked that the `tests/cert_check` test works from the GitLab pipeline since I can see, in the log output of `t_gigadb`, that the hostname of my staging server is being echoed and that it's `Certificate expires in: 57 days`.

I approve this PR.

rija committed:
Feat: Backend for FUW, improved test harness, improved GitLab pipeline and infrastructure for AWS

This PR introduces improvement to the Gitlab Pipeline, making it more modular and suitable for deploying to cloud environments.
The test harness has been improved as well for robustness and consistency and greater coverage (with the addition of smoke tests).
The use of Ansible and Terraform have been improved and extended for flexibility, ease maintenance and integration with GitLab 
to allow deployment of staging and live instances of GigaDB. 
The infrastructure and codebase for File Upload Wizard is present (but not enabled).

rija committed:
Add curl and jq to Docker DinD

rija commented:
Wil close this pull request as ``fuw-cicd`` has now been merged to gigascience/gigadb-website's ``develop``.
Will recreate it as PR to gigascience/gigadb-website's ``develop``.
